File: cpp-linux/xfa/08_picture_clause_formatting/08_picture_clause_formatting.cpp
// 08_picture_clause_formatting -- C++ port of
// examples\delphi\xfa\08_picture_clause_formatting\08_picture_clause_formatting.dpr
//
// XFA "flavor tour" example 8 of 10 -- PICTURE-CLAUSE FORMATTING: a
// "Purchase Receipt" exercising real num{}/date{}/text{} <picture> patterns
// applied both to plain bound values and to a value produced by a FormCalc
// <calculate> script (GrandTotalField = Item1Price + Item2Price +
// Item3Price, then formatted num{zzz,zz9.99}), proving the
// calculate-then-format pipeline order.
//
// Same pipeline as every example in this tour:
//   pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
//   pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
//
// Packet files are pre-split (08_picture_clause_formatting.template.xml /
// .datasets.xml) -- no XML parsing needed, just raw bytes.
//
// Does not rebuild LumasPdf.dll -- links against wrappers\c\lumaspdf.h and
// loads the LumasPdf.dll copied next to this .exe.
#include <lumaspdf.h>
#include <cstdio>
#include <string>
#include <fstream>
#include <sstream>

static bool ReadFileBytes(const std::string& path, std::string& out) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return false;
    std::ostringstream ss;
    ss << f.rdbuf();
    out = ss.str();
    return true;
}

static std::string ExeDir(const char* argv0) {
    std::string s(argv0);
    size_t p = s.find_last_of("\\/");
    return p == std::string::npos ? "." : s.substr(0, p);
}

static int RenderExample(const std::string& templatePath, const std::string& datasetsPath,
                          const std::string& outPdfPath) {
    printf("=== %s + %s -> %s ===\n", templatePath.c_str(), datasetsPath.c_str(), outPdfPath.c_str());

    std::string templateBuf, datasetsBuf;
    if (!ReadFileBytes(templatePath, templateBuf)) {
        printf("FILE-NOT-FOUND: %s\n", templatePath.c_str());
        return -100;
    }
    bool haveDatasets = ReadFileBytes(datasetsPath, datasetsBuf);
    printf("template packet bytes: %zu\n", templateBuf.size());
    printf("datasets packet bytes: %zu\n", haveDatasets ? datasetsBuf.size() : 0);

    PPDF pdf = pdfNewPDF();
    if (!pdf) {
        printf("pdfNewPDF FAILED\n");
        return -100;
    }

    int result = -100;
    if (!pdfCreateNewPDFA(pdf, outPdfPath.c_str())) {
        printf("pdfCreateNewPDFA FAILED\n");
        pdfDeletePDF(pdf);
        return -100;
    }

    int idx = pdfCreateXFAStreamA(pdf, "template", (void*)templateBuf.data(), (UI32)templateBuf.size());
    printf("pdfCreateXFAStreamA(template) -> index %d\n", idx);
    if (idx < 0) {
        printf("pdfCreateXFAStreamA(template) FAILED\n");
        pdfDeletePDF(pdf);
        return -100;
    }

    if (haveDatasets && !datasetsBuf.empty()) {
        idx = pdfCreateXFAStreamA(pdf, "datasets", (void*)datasetsBuf.data(), (UI32)datasetsBuf.size());
        printf("pdfCreateXFAStreamA(datasets) -> index %d\n", idx);
        if (idx < 0) {
            printf("pdfCreateXFAStreamA(datasets) FAILED\n");
            pdfDeletePDF(pdf);
            return -100;
        }
    } else {
        printf("(no datasets packet found -- template-only render)\n");
    }

    result = pdfRenderXFAForm(pdf);
    printf("pdfRenderXFAForm -> %d (expected: page count >= 1)\n", result);
    if (result < 1) {
        printf("RENDER-FAILED, code %d\n", result);
        pdfDeletePDF(pdf);
        return result;
    }

    if (!pdfCloseFile(pdf)) {
        printf("pdfCloseFile FAILED\n");
        pdfDeletePDF(pdf);
        return -101;
    }
    printf("OK: wrote %s\n", outPdfPath.c_str());
    pdfDeletePDF(pdf);
    return result;
}

int main(int argc, char** argv) {
    std::string dir = ExeDir(argv[0]);
    int r = RenderExample(
        dir + "/08_picture_clause_formatting.template.xml",
        dir + "/08_picture_clause_formatting.datasets.xml",
        dir + "/08_picture_clause_formatting.pdf");
    printf("RESULT|08_picture_clause_formatting=%d\n", r);

    if (r >= 1) {
        printf("\nExpect RESULT|08_picture_clause_formatting=1 (one page rendered).\n");
        printf("Verify with any text extractor (e.g. pypdf) -- the extracted text should read:\n");
        printf("  Purchase Receipt -- Picture-Clause Formatting\n");
        printf("  Customer: Acme Corp\n");
        printf("  Unit Price: 1,875.50\n");
        printf("  Discount: ($125.00)\n");
        printf("  Date: July 24, 2026\n");
        printf("  Phone: 555-123-4567\n");
        printf("  845.25 620.00 410.25\n");
        printf("  Grand Total: 1,875.50   (calculate result Item1+Item2+Item3, then formatted)\n");
    }
    return r >= 1 ? 0 : 1;
}

File: vbnet/xfa/08_picture_clause_formatting/08_picture_clause_formatting.vb
' 08_picture_clause_formatting -- VB.NET port of examples\delphi\xfa\08_picture_clause_formatting
'
' "Flavor tour" example 8 of 10: PICTURE-CLAUSE FORMATTING. Demonstrates the
' XFA <format><picture> formatter: real num{}/date{}/text{} picture patterns
' applied both to plain bound data values and to a value produced by a
' FormCalc <calculate> script, proving the calculate-then-format pipeline
' order (the picture clause is applied to the calc script's RESULT, not
' skipped for calculated fields). The form is a one-page "Purchase Receipt".
'
' Renders 08_picture_clause_formatting.template.xml / .datasets.xml
' (pre-split packet bytes) through the real LumasPdf.dll:
'
'   pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
'   pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
'
' Does NOT rebuild LumasPdf.dll -- links only against the generated
' wrappers\vbnet\LumasPdf.vb P/Invoke binding (LumasPdf.VB.dll) and the
' already-built LumasPdf.dll copied next to this exe.
Imports System
Imports System.IO
Imports System.Runtime.InteropServices
Imports LumasPdfSdk

Module PictureClauseFormatting

    Function RenderExample(ByVal TemplatePath As String, ByVal DatasetsPath As String, ByVal OutPdfPath As String) As Integer
        RenderExample = -100
        Console.WriteLine("=== " & Path.GetFileName(TemplatePath) & " -> " & Path.GetFileName(OutPdfPath) & " ===")
        If Not File.Exists(TemplatePath) Then
            Console.WriteLine("FILE-NOT-FOUND: " & TemplatePath)
            Return RenderExample
        End If

        Dim TemplateBuf() As Byte = File.ReadAllBytes(TemplatePath)
        Dim DatasetsBuf() As Byte = If(File.Exists(DatasetsPath), File.ReadAllBytes(DatasetsPath), New Byte() {})
        Console.WriteLine("template packet bytes: " & TemplateBuf.Length)
        Console.WriteLine("datasets packet bytes: " & DatasetsBuf.Length)

        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        If pdf = IntPtr.Zero Then
            Console.WriteLine("pdfNewPDF FAILED")
            Return RenderExample
        End If

        Try
            If Not LumasPdf.pdfCreateNewPDFW(pdf, OutPdfPath) Then
                Console.WriteLine("pdfCreateNewPDFA FAILED")
                Return RenderExample
            End If

            Dim TplHandle As GCHandle = GCHandle.Alloc(TemplateBuf, GCHandleType.Pinned)
            Dim Idx As Integer
            Try
                Idx = LumasPdf.pdfCreateXFAStreamW(pdf, "template", TplHandle.AddrOfPinnedObject(), CUInt(TemplateBuf.Length))
            Finally
                TplHandle.Free()
            End Try
            Console.WriteLine("pdfCreateXFAStreamA(template) -> index " & Idx)
            If Idx < 0 Then
                Console.WriteLine("pdfCreateXFAStreamA(template) FAILED")
                Return RenderExample
            End If

            If DatasetsBuf.Length > 0 Then
                Dim DsHandle As GCHandle = GCHandle.Alloc(DatasetsBuf, GCHandleType.Pinned)
                Try
                    Idx = LumasPdf.pdfCreateXFAStreamW(pdf, "datasets", DsHandle.AddrOfPinnedObject(), CUInt(DatasetsBuf.Length))
                Finally
                    DsHandle.Free()
                End Try
                Console.WriteLine("pdfCreateXFAStreamA(datasets) -> index " & Idx)
                If Idx < 0 Then
                    Console.WriteLine("pdfCreateXFAStreamA(datasets) FAILED")
                    Return RenderExample
                End If
            Else
                Console.WriteLine("(no datasets packet found -- template-only render)")
            End If

            Dim Rc As Integer = LumasPdf.pdfRenderXFAForm(pdf)
            Console.WriteLine("pdfRenderXFAForm -> " & Rc)
            If Rc < 1 Then
                Console.WriteLine("RENDER-FAILED, code " & Rc)
                Return RenderExample
            End If

            If Not LumasPdf.pdfCloseFile(pdf) Then
                Console.WriteLine("pdfCloseFile FAILED")
                RenderExample = -101
                Return RenderExample
            End If
            Console.WriteLine("OK: wrote " & OutPdfPath)
            RenderExample = Rc
        Finally
            LumasPdf.pdfDeletePDF(pdf)
        End Try
    End Function

    Sub Main()
        Dim ExeDir As String = AppDomain.CurrentDomain.BaseDirectory
        Dim R1 As Integer = RenderExample(
            Path.Combine(ExeDir, "08_picture_clause_formatting.template.xml"),
            Path.Combine(ExeDir, "08_picture_clause_formatting.datasets.xml"),
            Path.Combine(ExeDir, "08_picture_clause_formatting.pdf"))
        Console.WriteLine("RESULT|08_picture_clause_formatting=" & R1)
        If R1 >= 1 Then
            Console.WriteLine("Expected rendered text (verify with a PDF text extractor, e.g. pypdf):")
            Console.WriteLine("  Purchase Receipt -- Picture-Clause Formatting")
            Console.WriteLine("  Customer: Acme Corp")
            Console.WriteLine("  Unit Price: 1,875.50          (num{zzz,zz9.99})")
            Console.WriteLine("  Discount: ($125.00)           (num{($zzz,zz9.99)}, negative)")
            Console.WriteLine("  Date: July 24, 2026           (date{MMMM DD, YYYY})")
            Console.WriteLine("  Phone: 555-123-4567           (text{999-999-9999})")
            Console.WriteLine("  845.25 620.00 410.25          (raw calc inputs)")
            Console.WriteLine("  Grand Total: 1,875.50         (calculate result, THEN picture-formatted)")
        End If
    End Sub

End Module
